### PR TITLE
bpf: use FIB_LOOKUP_DEV_IFINDEX for fib_lookup

### PR DESCRIPTION
**Pull request by hzhou8**
When LBDevInheritIPAddr is configured, it results in datapath fib_lookup failure because the ifindex used for fib_lookup is still the native XDP device, which could be a L2 device without routing table. This PR solves the problem by passing the ifindex of the LBDevInheritIPAddr to datapath using a new macro FIB_LOOKUP_DEV_IFINDEX, and using it for fib_lookup, which keeping the DIRECT_ROUTING_DEV_IFINDEX for packet forwarding.

The PR is based on https://github.com/cilium/cilium/pull/14649

### CODE REVIEW

File: bpf/lib/nodeport.h [hunk 1]
```diff
@@ -449,7 +449,7 @@ int tail_nodeport_ipv6_dsr(struct __ctx_buff *ctx)
 		}
 	}
 
-	return ctx_redirect(ctx, fib_params.l.ifindex, 0);
+	return ctx_redirect(ctx, DIRECT_ROUTING_DEV_IFINDEX, 0);
```
Comment (brb) [inline bpf/lib/nodeport.h:452]: Why is this change required?

Comment (hzhou8) [reply bpf/lib/nodeport.h:452]: @brb For example, when there is a bridge br0 with eth0 attached, IP is set on br0 (for fib_lookup), and eth0 is a L2 device without IP, we should use br0's ifindex for fib_lookup, but still forward packets use eth0. In this case DIRECT_ROUTING_DEV_IFINDEX = eth0, FIB_LOOKUP_DEV_IFINDEX = br0.

Comment (brb) [reply bpf/lib/nodeport.h:452]: Sorry for taking long on this, and thanks for the explanation.

Anyway, I'm thinking maybe for the sake of the datapath simplicity, we could omit the `BPF_FIB_LOOKUP_DIRECT` flag in the `fib_lookup()`, and set ifindex=0. The downside of this approach is that might introduce a performance penalty. Would you mind measuring it?

If the suggested approach is OK, another thing is that for the DSR mode we don't actually need the LB IP addr in the datapath. So, maybe we could even get rid of the `--bpf-lb-dev-ip-addr-inherit` flag which would simplify the code by a bit.

Comment (hzhou8) [reply bpf/lib/nodeport.h:452]: @brb in fact I tried omitting the BPF_FIB_LOOKUP_DIRECT flag and setting ifindex = 0. However, it didn't work because the kernel helper function fib_lookup() doesn't support ifindex = 0. It validates that ifindex if valid and can get dev from it. I think the kernel helper should be fixed to avoid this validation when BPF_FIB_LOOKUP_DIRECT is not used. Until the kernel helper is fixed, this patch seems to be the only way to make it work.

For --bpf-lb-dev-ip-addr-inherit, I think the LB IP is still needed for DSR mode probably for environments where source IP range is checked in routers with RPF settings.

**Comment by joestringer** [comment]
@brb @hzhou8 are next steps on this PR clear? At the very least it seems to need a rebase.

**Comment by brb** [comment]
> are next steps on this PR clear? At the very least it seems to need a rebase.

I'm going to fix the kernel helpers. Afterwards, we will need to revive the discussion. Converted this PR to the draft for now.

**Comment by joestringer** [comment]
We believe that this issue has been fixed, can you try again with v1.12.0?
